src/ts/Document.ts

```diff
@@ -0,0 +1,223 @@
+
+	public _id?: string;
+
+	public _raw?: MvRecord;
+
+	/** Array of any errors which occurred during transformation from the database */
+	public transformationErrors: TransformDataError[];
+
+	/** Schema instance which defined this document */
+	private readonly schema: Schema | null;
+
+	/** Record array of multivalue data */
+	private record: MvRecord;
+
+	/** Indicates whether this document is a subdocument of a composing parent */
+	private readonly isSubdocument: boolean;
+
+	public constructor(schema: Schema, options: DocumentConstructorOptions);
+	public constructor(schema: null, options: DocumentConstructorOptionsRaw);
+	public constructor(schema: Schema | null, options: DocumentConstructorOptions) {
+		const { data, isSubdocument, record } = this.parseConstructorOptions(options);
+
+		this.schema = schema;
+		this.record = [];
+		this.isSubdocument = isSubdocument;
+		this.transformationErrors = [];
+
+		Object.defineProperties(this, {
+			schema: DEFAULT_PROPERTY_DESCRIPTOR,
+			record: {
+				...DEFAULT_PROPERTY_DESCRIPTOR,
+				writable: true,
+			},
+			isSubdocument: DEFAULT_PROPERTY_DESCRIPTOR,
+			transformationErrors: DEFAULT_PROPERTY_DESCRIPTOR,
+		});
+
+		if (record != null) {
+			this.transformRecordToDocument(record);
+		}
+		// load the data passed to constructor into document instance
+		assignIn(this, data);
+	}
+
+	/** Transform document structure to multivalue array structure */
+	public transformDocumentToRecord(): MvRecord {
+		return this.schema === null
+			? getIn(this, '_raw', [])
+			: Object.entries(this.schema.paths).reduce(
+					(record, [keyPath, schemaType]) => {
+						const value = getIn(this, keyPath, null);
+						return schemaType.set(record, schemaType.cast(value));
+					},
+					this.isSubdocument ? [] : cloneDeep(this.record),
+			  );
+	}
+
+	/** Build a list of foreign key definitions to be used by the database for foreign key validation */
+	public buildForeignKeyDefinitions(): BuildForeignKeyDefinitionsResult[] {
+		if (this.schema === null) {
+			return [];
+		}
+
+		// U2 does not allow commas in filenames so we can use it to separate filename/entityName combinations
+		const separator = ',';
+		const definitionMap = Object.entries(this.schema.paths).reduce(
+			(foreignKeyDefinitions, [keyPath, schemaType]) => {
+				const value = getIn(this, keyPath, null);
+				const definitions = schemaType.transformForeignKeyDefinitionsToDb(schemaType.cast(value));
+				// Deduplicate foreign key definitions by using a filename / entity name combination
+				// We could deduplicate using just the filename but ignoring the entity name could result in confusing error messages
+				definitions.forEach(({ filename, entityId, entityName }) => {
+					const key = `${filename}${separator}${entityName}`;
+					const accumulatedEntityIds = foreignKeyDefinitions.get(key) ?? new Set<string>();
+					// For array types we may need to validate multiple foreign keys
+					accumulatedEntityIds.add(entityId);
```

Comment:
There was an issue here where `Set.add` was being called with a spread against an array.  That is not valid syntax so I addressed it here.